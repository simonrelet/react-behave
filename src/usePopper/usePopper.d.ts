import { Modifiers, Placement, ReferenceObject } from 'popper.js'

export type UsePopperOptions = {
  arrow?: string | HTMLElement | null
  placement?: Placement
  modifiers?: Modifiers
  eventsEnabled?: boolean
  positionFixed?: boolean
}

export type UsePopperReturnValue = {
  style: CSSStyleDeclaration
  arrowStyle: CSSStyleDeclaration
  placement: Placement | null
  outOfBoundaries: boolean
  scheduleUpdate(): void
}

export function usePopper(
  reference: HTMLElement | ReferenceObject | null,
  popper: HTMLElement | null,
  options?: UsePopperOptions,
): UsePopperReturnValue
