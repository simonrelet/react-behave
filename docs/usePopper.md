# usePopper

A React hook around [Popper.js](https://popper.js.org).

## Usage

```js
import React from 'react'
import { usePopper } from 'react-behave'

function DropdownButton() {
  // Use states to manage the HTML elements to trigger updates and make sure
  // `usePopper` is up to date.
  const [reference, setReference] = React.useState(null)
  const [popper, setPopper] = React.useState(null)

  const [open, setOpen] = React.useState(false)

  const { style } = usePopper(reference, popper)

  return (
    <>
      <button ref={setReference} onClick={() => setOpen(open => !open)}>
        Open dropdown
      </button>

      {open && (
        <div ref={setPopper} style={style}>
          Popper element
        </div>
      )}
    </>
  )
}
```

## Parameters

<table>
  <thead>
    <tr>
      <th>Name</th>
      <th>Description</th>
    </tr>
  </thead>
  
  <tbody>
    <tr>
      <td><code>reference</code></td>
      <td>
        <strong>HTMLElement | <a href="https://popper.js.org/popper-documentation.html#referenceObject">ReferenceObject</a> | null</strong>
        <p>
          The reference element used to position the popper.
          See more in <a href="https://popper.js.org/popper-documentation.html#new_Popper_new">Popper.js's documentation</a>.
        </p>
      </td>
    </tr>
    <tr>
      <td><code>popper</code></td>
      <td>
        <strong>HTMLElement | null</strong>
        <p>
          The HTML element used as the popper.
          See more in <a href="https://popper.js.org/popper-documentation.html#new_Popper_new">Popper.js's documentation</a>.
        </p>
      </td>
    </tr>
    <tr>
      <td><code>options?</code></td>
      <td>
        <strong>object</strong> <em>= <code>{}</code></em>
        <p>The hook options.</p>
      </td>
    </tr>
    <tr>
      <td><code>options?.arrow?</code></td>
      <td>
        <strong>string | HTMLElement | null</strong>
        <p>
          Selector or node used as arrow.
          See more in <a href="https://popper.js.org/popper-documentation.html#modifiers..arrow.element">Popper.js's documentation</a>.
        </p>
      </td>
    </tr>
    <tr>
      <td><code>options?.placement?</code></td>
      <td>
        <strong><a href="https://popper.js.org/popper-documentation.html#Popper.placements">Placement</a></strong> <em>= <code>'bottom'</code></em>
        <p>
          Popperâ€™s placement.
          See more in <a href="https://popper.js.org/popper-documentation.html#Popper.Defaults.placement">Popper.js's documentation</a>.
        </p>
      </td>
    </tr>
    <tr>
      <td><code>options?.modifiers?</code></td>
      <td>
        <strong><a href="https://popper.js.org/popper-documentation.html#modifiers">Modifiers</a></strong> <em>= <code>{}</code></em>
        <p>
          List of modifiers used to modify the offsets before they are applied
          to the popper.
          They provide most of the functionalities of Popper.js.
          See more in <a href="https://popper.js.org/popper-documentation.html#Popper.Defaults.modifiers">Popper.js's documentation</a>.
        </p>
      </td>
    </tr>
    <tr>
      <td><code>options?.eventsEnabled?</code></td>
      <td>
        <strong>boolean</strong> <em>= <code>true</code></em>
        <p>
          Whether events (resize, scroll) are initially enabled.
          See more in <a href="https://popper.js.org/popper-documentation.html#Popper.Defaults.eventsEnabled">Popper.js's documentation</a>.
        </p>
      </td>
    </tr>
    <tr>
      <td><code>options?.positionFixed?</code></td>
      <td>
        <strong>boolean</strong> <em>= <code>false</code></em>
        <p>
          Set this to <code>true</code> if you want popper to position it self
          in "fixed" mode.
          See more in <a href="https://popper.js.org/popper-documentation.html#Popper.Defaults.positionFixed">Popper.js's documentation</a>.
        </p>
      </td>
    </tr>
  </tbody>
</table>

## Return value

**object**

The current state of the popper.
This object contains the following properties:

<table>
  <thead>
    <tr>
      <th>Name</th>
      <th>Description</th>
    </tr>
  </thead>
  
  <tbody>
    <tr>
      <td><code>style</code></td>
      <td>
        <strong>object</strong>
        <p>The CSS property to apply to the popper.</p>
      </td>
    </tr>
    <tr>
      <td><code>arrowStyle</code></td>
      <td>
        <strong>object</strong>
        <p>The CSS property to apply to the popper arrow.</p>
      </td>
    </tr>
    <tr>
      <td><code>placement</code></td>
      <td>
        <strong><a href="https://popper.js.org/popper-documentation.html#Popper.placements">Placement</a> | null</strong>
        <p>Placement applied to the popper.</p>
      </td>
    </tr>
    <tr>
      <td><code>outOfBoundaries</code></td>
      <td>
        <strong>boolean</strong>
        <p>
          <code>true</code> if the reference element is out of boundaries,
          useful to know when to hide the popper.
        </p>
      </td>
    </tr>
    <tr>
      <td><code>scheduleUpdate</code></td>
      <td>
        <strong>() => void</strong>
        <p>
          Schedules an update.
          It will run on the next UI update available.
        </p>
      </td>
    </tr>
  </tbody>
</table>
