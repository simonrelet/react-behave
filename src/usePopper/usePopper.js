import isEqual from 'lodash.isequal'
import PopperJS from 'popper.js'
import React from 'react'

const INITIAL_STATE = {
  arrowStyle: {},
  outOfBoundaries: false,
  placement: null,
  style: {
    position: 'absolute',
    top: 0,
    left: 0,
    opacity: 0,
    pointerEvents: 'none',
  },
}

function getCurrentValue(value, newValue) {
  return isEqual(value, newValue) ? value : newValue
}

function updateState(newState) {
  return state => {
    return {
      // Preserve the references if the objects are deeply equal.
      style: getCurrentValue(state.style, newState.styles),
      arrowStyle: getCurrentValue(state.arrowStyle, newState.arrowStyles),

      outOfBoundaries: newState.hide,
      placement: newState.placement,
    }
  }
}

const DEFAULT_MODIFIERS = {}

usePopper.PLACEMENTS = PopperJS.placements

export function usePopper(
  referenceRef,
  popperRef,
  {
    disabled = false,
    arrowRef,
    placement = 'bottom',
    modifiers = DEFAULT_MODIFIERS,
    eventsEnabled = true,
    positionFixed = false,
  } = {},
) {
  const popperJSRef = React.useRef(null)
  const [state, setState] = React.useState(INITIAL_STATE)

  React.useEffect(
    () => {
      if (
        !disabled &&
        referenceRef.current != null &&
        popperRef.current != null
      ) {
        popperJSRef.current = new PopperJS(
          referenceRef.current,
          popperRef.current,
          {
            placement,
            eventsEnabled,
            positionFixed,
            modifiers: {
              ...modifiers,
              arrow: {
                enabled: arrowRef != null && arrowRef.current != null,
                element: arrowRef == null ? null : arrowRef.current,
                ...modifiers.arrow,
              },
              applyStyle: {
                // We apply the styles ourselves.
                enabled: false,
              },
              updateStateModifier: {
                enabled: true,
                // The same as `applyStyle`.
                // https://popper.js.org/popper-documentation.html#modifiers..applyStyle.order
                order: 900,
                fn: data => {
                  // The popper element might have lost his parent during the
                  // computation of the position.
                  // It can be due to a fast removal of the popper.
                  // In case it happenned we don't dispatch the computed state.
                  if (data.instance.popper.parentElement != null) {
                    setState(updateState(data))
                  }

                  return data
                },
              },
            },
          },
        )

        return () => {
          setState(INITIAL_STATE)
          popperJSRef.current.destroy()
          popperJSRef.current = null
        }
      }
    },

    // 'arrowRef', 'popperRef', and 'referenceRef' are React Ref objects.
    // eslint-disable-next-line react-hooks/exhaustive-deps
    [disabled, placement, modifiers, eventsEnabled, positionFixed],
  )

  const scheduleUpdate = React.useCallback(() => {
    if (popperJSRef.current != null) {
      popperJSRef.current.scheduleUpdate()
    }
  }, [])

  return { ...state, scheduleUpdate }
}
