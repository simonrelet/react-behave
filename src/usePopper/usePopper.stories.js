import { select, text } from '@storybook/addon-knobs'
import { storiesOf } from '@storybook/react'
import React from 'react'
import { scrollableDecorator } from '../../.storybook/decorators'
import { usePopper } from './usePopper'

const stories = storiesOf('usePopper', module)
stories.addDecorator(scrollableDecorator)

stories.add('Place a popper next to a reference', () => (
  <PopperStory
    placement={select('placement', usePopper.PLACEMENTS, 'bottom')}
    referenceText={text('Reference text', 'Reference')}
    popperText={text('Popper text', 'Popper')}
  />
))

function PopperStory({ placement, referenceText, popperText }) {
  const referenceRef = React.useRef(null)
  const popperRef = React.useRef(null)

  const { style, scheduleUpdate } = usePopper(referenceRef, popperRef, {
    placement,
  })

  React.useLayoutEffect(() => {
    scheduleUpdate()
  }, [scheduleUpdate, referenceText, popperText])

  return (
    <>
      <Reference ref={referenceRef}>{referenceText}</Reference>
      <Popper ref={popperRef} style={style}>
        {popperText}
      </Popper>
    </>
  )
}

const Reference = React.forwardRef((props, ref) => {
  return <Block ref={ref} style={{ backgroundColor: '#8bc34a' }} {...props} />
})

const Popper = React.forwardRef(({ style, ...rest }, ref) => {
  return (
    <Block
      ref={ref}
      style={{ backgroundColor: '#ffeb3b', ...style }}
      {...rest}
    />
  )
})

const Block = React.forwardRef(({ style, ...props }, ref) => {
  return (
    <div
      ref={ref}
      style={{
        ...style,
        display: 'flex',
        alignItems: 'center',
        justifyContent: 'center',
        padding: 16,
      }}
      {...props}
    />
  )
})

stories.add('Toggle popper', () => (
  <TogglePopperStory
    placement={select('placement', usePopper.PLACEMENTS, 'bottom')}
    referenceText={text('Reference text', 'Open popper')}
    popperText={text('Popper text', 'Popper')}
  />
))

function TogglePopperStory({ placement, referenceText, popperText }) {
  const referenceRef = React.useRef(null)
  const popperRef = React.useRef(null)
  const [opened, setOpened] = React.useState(false)
  const { style, scheduleUpdate } = usePopper(referenceRef, popperRef, {
    disabled: !opened,
    placement,
  })

  React.useLayoutEffect(() => {
    scheduleUpdate()
  }, [scheduleUpdate, referenceText, popperText])

  return (
    <>
      <button ref={referenceRef} onClick={() => setOpened(opened => !opened)}>
        {referenceText}
      </button>

      {opened && (
        <div ref={popperRef} style={style}>
          {popperText}
        </div>
      )}
    </>
  )
}
