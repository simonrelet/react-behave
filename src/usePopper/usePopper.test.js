import { render } from '@testing-library/react'
import PopperJS from 'popper.js'
import React from 'react'
import { usePopper } from './usePopper'

function renderHook({ reference, popper, options }) {
  function HookedComponent({ renderHook }) {
    renderHook()
    return null
  }

  const hookResult = { current: null }

  const { rerender, ...result } = render(
    <HookedComponent
      renderHook={() => {
        hookResult.current = usePopper(reference, popper, options)
      }}
    />,
  )

  const prevReference = reference
  const prevPopper = popper
  const prevOptions = options

  return {
    ...result,
    hookResult,
    rerender: ({
      reference = prevReference,
      popper = prevPopper,
      options = prevOptions,
    }) => {
      rerender(
        <HookedComponent
          renderHook={() => {
            hookResult.current = usePopper(reference, popper, options)
          }}
        />,
      )
    },
  }
}

const STYLE = { top: 21, left: 42 }

function expectHookResult(
  hookResult,
  {
    arrowStyle = {},
    outOfBoundaries = false,
    placement = null,
    style = {
      position: 'absolute',
      top: 0,
      left: 0,
      opacity: 0,
      pointerEvents: 'none',
    },
  } = {},
) {
  expect(hookResult.current).toEqual({
    arrowStyle,
    outOfBoundaries,
    placement,
    style,
    scheduleUpdate: expect.any(Function),
  })
}

describe('usePopper', () => {
  afterEach(() => {
    PopperJS.mockReset()
  })

  it('should return the initial state when the reference is null', () => {
    const { hookResult } = renderHook({
      reference: { current: null },
      popper: { current: 'popper' },
    })

    expect(PopperJS.mock.calls).toHaveLength(0)
    expectHookResult(hookResult)
  })

  it('should return the initial state when the popper is null', () => {
    const { hookResult } = renderHook({
      reference: { current: 'reference' },
      popper: { current: null },
    })

    expect(PopperJS.mock.calls).toHaveLength(0)
    expectHookResult(hookResult)
  })

  it('should return initial state when the disabled option is true', () => {
    const { hookResult } = renderHook({
      reference: { current: 'reference' },
      popper: { current: 'popper' },
      options: { disabled: true },
    })

    expect(PopperJS.mock.calls).toHaveLength(0)
    expectHookResult(hookResult)
  })

  it('should return the computed state', async () => {
    const { hookResult } = renderHook({
      reference: { current: 'reference' },
      popper: { current: 'popper' },
    })

    expect(PopperJS.mock.calls).toMatchSnapshot()

    await PopperJS.mock.lastInstance.mockUpdateStateModifier({ styles: STYLE })
    expectHookResult(hookResult, { style: STYLE, placement: 'bottom' })
  })

  it('should cleanup on unmount', async () => {
    const { hookResult, unmount } = renderHook({
      reference: { current: 'reference' },
      popper: { current: 'popper' },
    })

    await PopperJS.mock.lastInstance.mockUpdateStateModifier({ styles: STYLE })
    expectHookResult(hookResult, { style: STYLE, placement: 'bottom' })

    unmount()
    expect(PopperJS.mock.lastInstance.destroy).toHaveBeenCalled()
  })

  // should include the arrow in the computed state
  it('should include the arrow in the computed state', async () => {
    const { hookResult } = renderHook({
      reference: { current: 'reference' },
      popper: { current: 'popper' },
      options: {
        arrowRef: { current: 'arrow' },
      },
    })

    expect(PopperJS.mock.calls).toMatchSnapshot()

    await PopperJS.mock.lastInstance.mockUpdateStateModifier({
      styles: STYLE,
      arrowStyles: STYLE,
    })

    expectHookResult(hookResult, {
      style: STYLE,
      arrowStyle: STYLE,
      placement: 'bottom',
    })
  })

  it("should keep the same style references of computed state when they haven't changed", async () => {
    const { hookResult } = renderHook({
      reference: { current: 'reference' },
      popper: { current: 'popper' },
    })

    await PopperJS.mock.lastInstance.mockUpdateStateModifier({
      styles: STYLE,
      arrowStyles: STYLE,
    })

    expectHookResult(hookResult, {
      style: STYLE,
      arrowStyle: STYLE,
      placement: 'bottom',
    })

    const currentResult = hookResult.current

    await PopperJS.mock.lastInstance.mockUpdateStateModifier({
      styles: { ...STYLE },
      arrowStyles: { ...STYLE },
    })

    expect(hookResult.current.style).toBe(currentResult.style)
    expect(hookResult.current.arrowStyles).toBe(currentResult.arrowStyles)
  })

  // should ignore calls to scheduleUpdate when there are no PopperJS instances
  it('should ignore calls to scheduleUpdate when there are no PopperJS instances', () => {
    const { hookResult } = renderHook({
      reference: { current: null },
      popper: { current: null },
    })

    // Don't throw.
    hookResult.current.scheduleUpdate()
  })
})
