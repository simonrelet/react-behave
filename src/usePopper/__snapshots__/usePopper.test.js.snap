// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`usePopper should include the arrow in the computed state 1`] = `
Array [
  Array [
    "reference",
    "popper",
    Object {
      "eventsEnabled": true,
      "modifiers": Object {
        "applyStyle": Object {
          "enabled": false,
        },
        "arrow": Object {
          "element": "arrow",
          "enabled": true,
        },
        "updateStateModifier": Object {
          "enabled": true,
          "fn": [Function],
          "order": 900,
        },
      },
      "placement": "bottom",
      "positionFixed": false,
    },
  ],
]
`;

exports[`usePopper should return the computed state 1`] = `
Array [
  Array [
    "reference",
    "popper",
    Object {
      "eventsEnabled": true,
      "modifiers": Object {
        "applyStyle": Object {
          "enabled": false,
        },
        "arrow": Object {
          "element": null,
          "enabled": false,
        },
        "updateStateModifier": Object {
          "enabled": true,
          "fn": [Function],
          "order": 900,
        },
      },
      "placement": "bottom",
      "positionFixed": false,
    },
  ],
]
`;
